analyzers/src/SonarAnalyzer.CSharp/Rules/CloudNative/AzureFunctionsReuseClients.cs

```diff
@@ -0,0 +1,101 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Extensions;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Wrappers;
+using StyleCop.Analyzers.Lightup;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class AzureFunctionsReuseClients : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S6420";
+        private const string MessageFormat = "Reuse client instances rather than creating new ones with each function invocation.";
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly ImmutableArray<KnownType> ReusableClients = ImmutableArray.Create(
+            KnownType.System_Net_Http_HttpClient,
+            // ComosDb. DocumentClient is superseeded by CosmosClient
+            KnownType.Microsoft_Azure_Documents_Client_DocumentClient,
+            KnownType.Microsoft_Azure_Cosmos_CosmosClient,
+            // Servicebus V5
+            KnownType.Microsoft_Azure_ServiceBus_Management_ManagementClient,
+            KnownType.Microsoft_Azure_ServiceBus_QueueClient,
+            KnownType.Microsoft_Azure_ServiceBus_SessionClient,
+            KnownType.Microsoft_Azure_ServiceBus_SubscriptionClient,
+            KnownType.Microsoft_Azure_ServiceBus_TopicClient,
+            // Servicebus V7
+            KnownType.Azure_Messaging_ServiceBus_ServiceBusClient,
+            KnownType.Azure_Messaging_ServiceBus_Administration_ServiceBusAdministrationClient,
+            // Storage
+            KnownType.Azure_Storage_Blobs_BlobServiceClient,
+            KnownType.Azure_Storage_Queues_QueueServiceClient,
+            KnownType.Azure_Storage_Files_Shares_ShareServiceClient,
+            KnownType.Azure_Storage_Files_DataLake_DataLakeServiceClient,
+            // Ressource manager
+            KnownType.Azure_ResourceManager_ArmClient);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (c.AzureFunctionMethod() is not null
+                    && CreatedResuableClient(c.SemanticModel, c.Node) is not null
```

Comment:
I like to express, that these functions are pure. The closest you can get in C# is to make the functions static and don't pass anything that may cause side effects. `c.ReportIssue` violates this. In analyzers that might this is maybe not so important, because `ReportIssue` is basically the only side effect you can cause.
I could also make model, node, and cancelationToken locals though to reduce repetition.

Reply:
We don't really care about pureness. I'd use `c` just to reduce the noise.

Reply:
I just realized that SyntaxNodeAnalysisContext is a struct. Passing it around makes even less sense with all the field copying (7 fields if I counted that right).

Reply:
That doesn't cost anything